src/sidebar/components/hooks/use-user-filter-options.js

```diff
@@ -0,0 +1,79 @@
+import { useMemo } from 'preact/hooks';
+
+import { useStoreProxy } from '../../store/use-store';
+import { username } from '../../util/account-id';
+
+/** @typedef {import('../../store/modules/filters').FilterOption} FilterOption */
+
+/**
+ * Generate a list of users for filtering annotations; update when set of
+ * annotations or filter state changes meaningfully.
+ *
+ * @return {FilterOption[]}
+ */
+export default function useUserFilterOptions() {
+  const store = useStoreProxy();
+  const annotations = store.allAnnotations();
+  const focusFilters = store.getFocusFilters();
+  const showDisplayNames = store.isFeatureEnabled('client_display_names');
+
+  return useMemo(() => {
+    // Determine unique users (authors) in annotation collection
+    const users = {};
+    annotations.forEach(annotation => {
+      // Display a username, not an account ID. If we're dealing with an
+      // account ID, we need to username() it
+      const _username = username(annotation.user)
+        ? username(annotation.user)
+        : annotation.user;
+      const displayValue =
+        showDisplayNames && annotation.user_info?.display_name
+          ? annotation.user_info.display_name
+          : _username;
+      users[annotation.user] = displayValue;
+    });
+
+    // If user-focus is configured (even if not applied) add a filter
```

Comment:
The comments in here were fantastically hard to write and I think it's still really confusing. This is both because my face is really close to the forest and because the underlying situation is smelly.

Reply:
To check my understanding here, the practical use case for this currently is that if the teacher is in grading mode in the LMS and has a particular user focused, when they open the notebook they should see that particular user, even if that user has not created any annotations?

Reply:
Its purpose is manifold:

1. Make sure that a focused user shows up in the list, even if they have authored no annotations in the current set
2. Make sure the focused user only shows up _once_ in the list (don't duplicate)
3. Make sure the user filter correctly shows the focused user as selected (this is the trickiest piece to explain)

The first requirement is fairly straightforward—we need to make sure the focused user is included in the list at all.

Perhaps an example is in order, as this is darned complicated to explain.

So let's talk about 2 and 3:

Let's say the app is configured to focus on a user, and the config for focused user provide a username and a display name for that user of `bumble` and `Bumble Bee`, respectively. Let's also say the Notebook is opened and the set of annotations contains several that were authored by `acct:bumble@flowertown` and a `display_name` (gleaned from one of their annotation objects)of `Bumble Q Bee`.

The hook will first generate a list of users that have authored at least one annotation—that would put `acct:bumble@flowertown / Bumble Q Bee` in the list of user filter options.

We need to make sure that we don't end up with two filter options for one person— `bumble / Bumble Bee` vs. `acct:bumble@flowertown / Bumble Q Bee`. A little bit of fiddling is required because the focus-user value could be either a username or an accountID. So that's part of this, eliminating dupes. This is where I got to initially and it satisfies item 2 above...but then I realized a wrinkle.

To satisfy item 3, we need to be sure that, in the case of the focused user already being in the users-with-annotations list, that filter option is superseded (replaced) by the focused-user filter variant. This matters when the focused-user filter value is a `username` instead of an `accountId`, as it is in our example.

Upon opening the Notebook in our example case,  there is an applied user filter (because of the focused-user config) for `bumble / Bumble Bee`. If we didn't replace the user filter option with the focus-filter variant—if we left it as `acct:bumble@flowertown / Bumble Q Bee`—the user filter would not correctly recognize that we are filtering on this user because the value of the applied (user-focus) filter is `bumble`, not `acct:bumble@flowertown`.

I'm making the assumption that the best UX for instructors in the grading mode would be to see annotations initially filtered by their focused user. I mean, we could just decide to not honor user-focus in Notebook, but I think it's nicer this way.

This whole screaming complexity could be averted if we only accepted `accountID`s for user-focus config.

Reply:
It being this hard to explain is more than a little bit of a design smell.